DPJTH-1 Add password loading support

datalogics-robl opened the pull request:
#### Changes in this pull request

Loads the password from a properties file. The file:
- is named `hsm.properties`
- contains an entry `hsm.password`
- can be located in either the current working directory or the user's home directory
#### Contributes to  [DPJTH-1](https://jira.datalogics.com/browse/DPJTH-1)
#### Checklist for approving this pull request

(PR creator amend this with more conditions if necessary)
- [x] There are **unit tests** for new/changed code, or a good explanation why this was not possible.
- [x] All **CI builders** have indicated success (Give them a few minutes to notice the pull request.)
- [x] The **Pull request title** has the JIRA issue numbers separated by spaces (if any), a space, and then a short, but descriptive summary.
- [x] **Commit messages** are well formed: [A note about Git commit messages](http://www.tpope.net/node/106)
- [x] New public packages, classes, and methods are **documented**. (Strongly consider documenting private classes and methods.)
- [x] Verify that copying files to the user's home directory will not cause problems on an automated build system like Jenkins (consult with @datalogics-kam)
#### Blockers

Add a checkbox for yourself with your **&#64;name** to this list if you are holding this PR open for review. PR Shepherd, please hold off merging this PR until these are all checked:
- [x] _&lt;add your name here with an **@**>_

Commit:
Move variable closer to its usage

Commit:
Depend on Apache Commons Configuration 2 for properties file support

This will let us store the password outside of the source code. Note that we
also need to pull in Apache Commons Beanutils to use some Configuration 2
features.

Commit:
Add ConfigurationUtils helper class

This class provides a Configuration object from a file on disk, whose name is
supplied by the client. This first version only searches the local directory
for the file.

Commit:
Modify the sample to obtain a password from a file in the working directory

This uses the ConfigurationUtils class to read the file 'hsm.properties' from
the current directory. This has the entry 'hsm.password', which will the
password to the HSM device.

Commit:
Tell Git to ignore .properties files

This guards against accidental check-ins of the HSM password.

Commit:
Have ConfigurationUtils look for properties files in the user's home directory

This sets up a search strategy for locating properties files.

Commit:
Refactor code to fit in our line length limit

datalogics-robl committed:
Use constants for properties file

datalogics-robl committed:
Throw IOException if configuration file could not be loaded

Wrap the ConfigurationException, which has a cryptic error message.

datalogics-robl committed:
Rename 'ConfigurationUtils' to 'SampleConfigurationUtils'

Apache Commons Configurations2 already has a 'ConfigurationUtils' object, and
while it is in a different package, using a distinct name cuts down on
confusion.

datalogics-robl committed:
Add a test.properties file for unit testing

datalogics-robl committed:
Add unit tests for the SampleConfigurationUtils class.

datalogics-robl committed:
Fix indentation errors in the POM

datalogics-kam commented on .gitignore:
```diff
@@ -66,6 +69,7 @@ dependency-reduced-pom.xml
 !.settings/org.eclipse.m2e.core.prefs
 !src/main/resources/com/datalogics/pdf/hsm/samples/pdfjavatoolkit-ds.pdf
 !user-documentation.pdf
+!src/test/resources/com/datalogics/pdf/hsm/samples/util/test.properties
```
Oooh, advanced ignoring.

datalogics-robl replied:
Actually, is ignoring all `.properties` files the smartest way to do this? We want to keep out the password file, but users might add other `.properties` files of their own that they want to keep in version control.

datalogics-kam replied:
You could ignore the specific `hsm.properties` file, which would be the kind that has the password.

datalogics-kam commented:
Can you put something in the README that reminds users about how to specify their HSM password?

datalogics-robl committed:
Add section to README describing how to add a password properties file

The demonstration password was generated with:

  dd if=/dev/random bs=1 count=16 | base64

datalogics-devauto commented:
Build failed.

datalogics-devauto commented:
Build failed.

datalogics-robl committed:
Redirect the JVM to use a fake user home directory

This avoids copying files to a user's actual home directory, which may be
protected on automated builders.  A directory in target/test-output is used
instead.

datalogics-devauto commented:
Build succeeded.

jrowden8 commented on src/test/java/com/datalogics/pdf/hsm/samples/util/SampleConfigurationUtilsTest.java:
```diff
@@ -0,0 +1,106 @@
+    private static void createFakeUserHomeDirectory() throws IOException {
+        final File fakeUserHomeDir = new File(new File(new File("target"), "test-output"), "fake-user-home-dir");
+        Files.createDirectories(fakeUserHomeDir.toPath());
+        System.setProperty("user.home", fakeUserHomeDir.getCanonicalPath());
```
Nice work around 👍

jrowden8 commented:
Great work here @datalogics-robl! Looks good.
